bpf/k_unix_sock.h

```diff
@@ -0,0 +1,298 @@
+    // threads in the runtime.
+    u64 sock_p = (u64)sk;
+    ensure_sent_event(id, &sock_p);
+
+    recv_args_t args = {
+        .sock_ptr = (u64)sk,
+    };
+
+    get_iovec_ctx((iovec_iter_ctx *)&args.iovec_ctx, msg);
+
+    bpf_map_update_elem(&active_recv_args, &id, &args, BPF_ANY);
+
+    return 0;
+}
+
+static __always_inline int return_unix_recvmsg(void *ctx, u64 id, int copied_len) {
+    recv_args_t *args = (recv_args_t *)bpf_map_lookup_elem(&active_recv_args, &id);
+
+    bpf_dbg_printk(
+        "=== return unix recvmsg id=%d args=%llx copied_len %d ===", id, args, copied_len);
+
+    if (!args || (copied_len <= 0)) {
+        return 0;
+    }
+
+    iovec_iter_ctx *iov_ctx = (iovec_iter_ctx *)&args->iovec_ctx;
+
+    if (!iov_ctx->iov && !iov_ctx->ubuf) {
+        bpf_dbg_printk("iovec_ptr found in kprobe is NULL, ignoring this tcp_recvmsg");
+        bpf_map_delete_elem(&active_recv_args, &id);
+
+        return 0;
+    }
+
+    struct sock *sock_ptr = (struct sock *)args->sock_ptr;
+
+    unsigned long inode_number;
+    unsigned long peer_inode_number = 0;
+    BPF_CORE_READ_INTO(&inode_number, sock_ptr, sk_socket, file, f_inode, i_ino);
+    struct unix_sock *usock = unix_sock_from_sk(sock_ptr);
+
+    if (usock) {
+        BPF_CORE_READ_INTO(&peer_inode_number, usock, peer, sk_socket, file, f_inode, i_ino);
+    }
+
+    bpf_dbg_printk("ino %d, peer ino %d", inode_number, peer_inode_number);
+
+    pid_connection_info_t p_conn = {.conn = {0}};
+
+    pid_connection_info_for_inode(id, &p_conn, inode_number, peer_inode_number);
+
+    bpf_map_delete_elem(&active_recv_args, &id);
+
+    u8 *buf = iovec_memory();
+    if (buf) {
+        copied_len = read_iovec_ctx(iov_ctx, buf, copied_len);
+        if (copied_len) {
+            // doesn't return must be logically last statement
+            handle_buf_with_connection(ctx, &p_conn, buf, copied_len, NO_SSL, TCP_RECV, 0);
+        } else {
+            bpf_dbg_printk("Not copied anything");
+        }
+    }
+
+    return 0;
+}
+
+SEC("kretprobe/unix_stream_recvmsg")
+int BPF_KRETPROBE(kretprobe_unix_stream_recvmsg, size_t copied) {
+    u64 id = bpf_get_current_pid_tgid();
+
+    if (!valid_pid(id)) {
+        return 0;
+    }
+
+    bpf_dbg_printk("=== kretprobe_unix_stream_recvmsg id=%d copied_len %d ===", id, copied);
+
+    return return_unix_recvmsg(ctx, id, copied);
+}
+
+SEC("kprobe/unix_stream_sendmsg")
+int BPF_KPROBE(kprobe_unix_stream_sendmsg, struct socket *sock, struct msghdr *msg, size_t size) {
+    u64 id = bpf_get_current_pid_tgid();
+
+    if (!valid_pid(id)) {
+        return 0;
+    }
+
+    bpf_printk("=== unix_stream sendmsg %d ===", id);
+
+    struct sock *sk;
+    BPF_CORE_READ_INTO(&sk, sock, sk);
+
+    unsigned long inode_number;
+    unsigned long peer_inode_number = 0;
+    BPF_CORE_READ_INTO(&inode_number, sock, sk, sk_socket, file, f_inode, i_ino);
+
+    struct unix_sock *usock = unix_sock_from_socket(sock);
```

Comment:
in this case, is this equivalent to:

```
struct unix_sock *usock = unix_sock_from_sk(sk);
```
?
If so, then that will save us a few extra reads

Reply:
Yup, another code iteration I guess, we don't need to re-read it.